### PR TITLE
AuthDAO deleteGroup and deleteWorkspaceAccessGroups

### PR DESCRIPTION
**Pull request by jmthibault79**
DSDEEPB-1800
- [x] **Submitter**: Rebase to develop. DO NOT SQUASH
- [x] **Submitter**: Make sure Swagger is updated if API changes
- [x] **Submitter**: Make sure documentation for code is complete
- [x] **Submitter**: Review code comments; remove done TODOs, create stories for remaining TODOs
- [x] **Submitter**: Include the JIRA issue number in the PR description
- [x] **Submitter**: Add description or comments on the PR explaining the hows/whys (if not obvious)
- [x] Tell  that the PR exists if he wants to look at it
- [x] Anoint a lead reviewer (LR). **Assign PR to LR**
- [x] **LR**: Initial review by LR and others.
- [x] Comment / review / update cycle:
  - Rest of team may comments on PR at will
  - **LR assigns to submitter** for feedback fixes
  - Submitter updates documentation as needed
  - Submitter rebases to develop again if necessary
  - Submitter makes further commits. DO NOT SQUASH. **Reassign to LR** for further feedback
- [x]  sign off
- [x] **LR** sign off
- [x] **Assign to submitter** to finalize
- [x] **Submitter**: Squash commits, rebase if necessary
- [x] **Submitter**: Verify all tests go green, including CI tests
- [x] **Submitter**: Merge to develop 
- [x] **Submitter**: Delete branch after merge
- [x] **Submitter**: Check configuration files in Jenkins in case they need changes
- [x] **Submitter**: Verify swagger UI on dev server still works after deployment
- [x] **Submitter**: Inform other teams of any API changes via hipchat and/or email
- [x] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

### CODE REVIEW

**Comment by helgridly** [comment]
You might want to take a look at GraphDAODeleteSpec, which also checks that workspaces don't leave vertices behind when they're deleted.

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphAuthDAO.scala [hunk 1]
```diff
@@ -52,6 +58,12 @@ class GraphAuthDAO extends AuthDAO with GraphDAO {
       WorkspaceAccessLevels.Read -> rGroup)
   }
 
+  override def deleteWorkspaceAccessGroups(workspace: Workspace, txn: RawlsTransaction) = {
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphAuthDAO.scala]: I don't think this function belongs here, you should probably just inline it

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphAuthDAO.scala [hunk 2]
```diff
@@ -25,6 +24,13 @@ class GraphAuthDAO extends AuthDAO with GraphDAO {
     rawlsGroup
   }
 
+  override def deleteGroup(rawlsGroup: RawlsGroupRef, txn: RawlsTransaction) = txn withGraph { db =>
+    val vertex = getGroupVertex(db, rawlsGroup.groupName) match {
+      case None => throw new RawlsException("Cannot delete group %s from database because it does not exist".format(rawlsGroup.groupName))
+      case Some(vertex) => removeObject(vertex, db)
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphAuthDAO.scala:30]: why not call graphDAO.removeObject?

File: src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala [hunk 3]
```diff
@@ -193,6 +184,7 @@ class WorkspaceService(userInfo: UserInfo, dataSource: DataSource, containerDAO:
         }
 
         gcsDAO.deleteBucket(userInfo, workspaceContext.workspace.workspaceId).map { _ =>
+          containerDAO.authDAO.deleteWorkspaceAccessGroups(workspaceContext.workspace, txn)
```
Comment (helgridly) [inline src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:187]: Should we also be deleting the googly group too?

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphAuthDAO.scala (hunk 2 above)
Comment (helgridly) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphAuthDAO.scala:30]: Isn't that exactly what he's doing?

File: src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala (hunk 3 above)
Comment (helgridly) [inline src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:187]: And at what point do we delete sub-groups and users in google-land? Never?

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphAuthDAO.scala (hunk 2 above)
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphAuthDAO.scala:30]: nevermind

File: src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala (hunk 3 above)
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:187]: Yes, delete the google group too.

I don't think there is a delete user use case. A group is a top level thing, a sub group is not owned by its container. So don't delete sub groups.

Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:187]: gcsDAO.deleteBucket might actually delete the groups which is  unexpected

Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:187]: createBucket and deleteBucket do in fact create and remove the Google Groups.

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphAuthDAO.scala (hunk 1 above)
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphAuthDAO.scala]: You mean push it down to AuthDAO?

**Comment by jmthibault79** [comment]
Good catch: I did leave a vertex behind.  Working on that now.

**Comment by jmthibault79** [comment]
Comments addressed @helgridly and @dvoet

File: src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala (hunk 3 above)
Comment (helgridly) [inline src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:187]: does this include deleting the owner/writer/reader groups? i'm not sure how the google side of things works.

edit:\ what i mean is that workspaces have the acl map attached to them, which is implemented as a separate RawlsGroup "belonging to" that workspace (whatever that means) for each access level. if they have parallels in google then they should go too

File: src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/AuthDAO.scala [hunk 4]
```diff
@@ -8,7 +8,15 @@ trait AuthDAO {
...
+  def deleteGroup(rawlsGroup: RawlsGroupRef, txn: RawlsTransaction)
+
   def createWorkspaceAccessGroups(workspaceName: WorkspaceName, userInfo: UserInfo, txn: RawlsTransaction): Map[WorkspaceAccessLevel, RawlsGroupRef]
 
+  def deleteWorkspaceAccessGroups(workspace: Workspace, txn: RawlsTransaction) = {
+    workspace.accessLevels foreach { case (_, group) =>
+      deleteGroup(group, txn)
+    }
+  }
+
```
Comment (helgridly) [inline src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/AuthDAO.scala:20]: Nice!

File: src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala (hunk 3 above)
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:187]: `deleteBucket` has an internal method `deleteGroups` which removes the Workspace's 3 ACL groups in google already.

Comment (helgridly) [inline src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:187]: So it does. I'm picking nits now, but I'd recommend replacing references to `groupAccessLevelsAscending` in that code with `WorkspaceAccessLevels.all` so they don't go out of sync if we add more access levels. If you're wary of requesting deletion of buckets that don't exist, at least move the definition of `groupAccessLevelsAscending` next to `all` so they're in the same place.

Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:187]: OK, did the latter.

**Comment by helgridly** [comment]
:+1:

**Comment by dvoet** [comment]
:+1:

**Pull request merged**